File: android/src/main/kotlin/com/gdelataillade/alarm/models/VolumeSettings.kt
package com.gdelataillade.alarm.models

import VolumeFadeStepWire
import VolumeSettingsWire
import kotlinx.serialization.Serializable
import kotlin.time.Duration
import kotlin.time.Duration.Companion.milliseconds

@Serializable
data class VolumeSettings(
    val volume: Double?,
    val fadeDuration: Duration?,
    val fadeSteps: List<VolumeFadeStep>,
    val volumeEnforced: Boolean
) {
    companion object {
        fun fromWire(e: VolumeSettingsWire): VolumeSettings {
            return VolumeSettings(
                e.volume,
                e.fadeDurationMillis?.milliseconds,
                e.fadeSteps.map { VolumeFadeStep.fromWire(it) },
                e.volumeEnforced,
            )
        }
    }
}

@Serializable
data class VolumeFadeStep(
    val time: Duration,
    val volume: Double
) {
    companion object {
        fun fromWire(e: VolumeFadeStepWire): VolumeFadeStep {
            return VolumeFadeStep(
                e.timeMillis.milliseconds,
                e.volume,
            )
        }
    }
}

File: ios/alarm/Sources/alarm/models/VolumeSettings.swift
import Foundation

struct VolumeSettings {
    var volume: Double?
    var fadeDuration: TimeInterval?
    var fadeSteps: [VolumeFadeStep]
    var volumeEnforced: Bool

    static func from(wire: VolumeSettingsWire) -> VolumeSettings {
        return VolumeSettings(
            volume: wire.volume,
            fadeDuration: wire.fadeDurationMillis.map { Double($0) / 1_000.0 },
            fadeSteps: wire.fadeSteps.map(VolumeFadeStep.from),
            volumeEnforced: wire.volumeEnforced
        )
    }
}

struct VolumeFadeStep {
    var time: TimeInterval
    var volume: Double

    static func from(wire: VolumeFadeStepWire) -> VolumeFadeStep {
        return VolumeFadeStep(
            time: Double(wire.timeMillis) / 1_000.0,
            volume: wire.volume
        )
    }
}
